Fix Terraform caching mishaps.

mcdonnnj opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description

This PR fixes two issues:
1. The Terraform initialization step in the lint job of the build workflow was attempting to initialize the `.terraform` directory because it is cached. This caused runs to fail.
1. The version of Terraform we were installing is not what the system was using. We symlinked to `/usr/bin/terraform`, but `/usr/local/bin` had higher precedence in the PATH of the virtual environment. This also caused builds to fail because of version conflicts.

These are remedied by:
1. Adding "guardrails" to the initialization step so that it does not attempt to initialize a `.terraform` directory.
1. Moving the `terraform` binary that is installed on GH Actions runners at `/usr/local/bin/terraform`, and then symlink our version pinned copy.

Once this PR is approved I will make the same updates to the workflows in [cisagov/skeleton-tf-module](https://github.com/cisagov/skeleton-tf-module) and [cisagov/skeleton-ansible-role-with-test-user](https://github.com/cisagov/skeleton-ansible-role-with-test-user) since they use Terraform in the same way.


### UPDATE

After looking at [cisagov/skeleton-tf-module](https://github.com/cisagov/skeleton-tf-module) and [cisagov/skeleton-ansible-role-with-test-user](https://github.com/cisagov/skeleton-ansible-role-with-test-user), I was reminded of the rationale behind why we didn't cache the `.terraform` directory previously. As a result, I am removing `.terraform` caching and adding an explanatory comment for why. I am leaving the guardrails in because it's good to have them in case Terraform's behavior changes in the future.

Terraform code describing behavior:
https://github.com/hashicorp/terraform/blob/83630a7003fb8b868a3bf940798326634c3c6acc/command/init.go#L491-L494

<!--- Describe your changes in detail -->

## 💭 Motivation and Context

@jsf9k noticed that this project, and those descended from it, were failing [apb](https://github.com/cisagov/action-apb) triggered builds. Upon investigating, he noticed that it was initializing the `.terraform` directory. He reached out to me and I began to look into the cause. I realized that the initialization step had been written when there was no caching of the `.terraform` directory, and so lacked guardrails to prevent the observed behavior. In the course of testing the fix, I noticed that the output of the initialization step referenced Terraform v0.13. Upon investigation I found that the runner had Terraform v0.13 pre-installed (this was added in the [20200817 build](https://github.com/actions/virtual-environments/releases/tag/ubuntu20%2F20200817.1) of the GH Actions virtual environments). This necessitated a change in the Terraform installation step of the same job to ensure our pinned version was being used.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## 🧪 Testing

GitHub Actions jobs successfully pass with caching still enabled. I also verified that we are using the pinned version of Terraform in our workflow now.

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

mcdonnnj committed:
Update Terraform initialization in GitHub Actions workflow

jsf9k noticed an issue with the lint job in the build workflow where with
caching of the .terraform directory, when the job initialized Terraform
directories it would try and initialize the .terraform directory. This would
cause the job to fail. This adds appropriate guardrails to avoid that issue.

mcdonnnj committed:
Add step to get Terraform information

mcdonnnj committed:
Add additional debugging information about Terraform version

mcdonnnj committed:
Add step to find all terraform versions

mcdonnnj committed:
Run binary finding step under sudo

mcdonnnj committed:
Symlink additional path to Terraform binary

mcdonnnj committed:
Pare down debugging information

mcdonnnj committed:
Move existing Terraform binary

Move the pre-installed binary so we can preserve it, but still allow us to
symlink our version pinned, manually installed version.

mcdonnnj committed:
Remove remaining debugging steps

jsf9k reviewed (approved):
Nice work!  Thanks for fixing this.

dav3r reviewed (approved):
Strong work! 💪 💼

mcdonnnj committed:
Adjust Terraform initialization step

Add output to show which path is being initialized and add the -upgrade flag.

mcdonnnj committed:
Remove .terraform caching and add explanatory comment

If we are caching the .terraform directory then we need to use the -upgrade
option to ensure the modules and plugins we use for linting are up-to-date.
However, the -upgrade=true option ignores what is already downloaded, and just
downloads the latest allowable versions. This defeats the purpose of caching,
so we will not bother.

dav3r reviewed (approved):
Good catch! 👍 👍

jsf9k reviewed (approved):
Still McLovin' this PR!
